### PR TITLE
Add promise support for onClose handlers. fixes #69

### LINKED ISSUE #69: async function as onClose handler
If I register an asynchronous onClose handler as a two argument (context, done) function:

```javascript
async function myPlugin (instance, opts) {
  instance.onClose((context, done) => {
    setTimeout(done, 3000)
  })
}
```
when app.close() is called, the next onClose handler does not start until the done() callback is not called on the previous one.

But if I register an asynchronous onClose handler as an async function:

```javascript
async function myPlugin (instance, opts) {
  instance.onClose(async () => {
    await new Promise(resolve => setTimeout(resolve, 3000))
  })
}
```
when app.close() is called, the next onClose handler does not wait until the async function finishes, i.e. the promise resolves.

Is it intentional? It would be much easier to write asynchronous onClose handlers with async functions.

### PR DESCRIPTION
**Pull request by gyszalai**

### CODE REVIEW

**Comment by gyszalai** [comment]
oops, I forgot the tests.

**Comment by gyszalai** [comment]
@delvedor please review the tests. I'm not sure if they are correct this way.

File: test/close.test.js [hunk 1]
```diff
@@ -404,3 +404,31 @@ test('close without a cb returns a promise when attaching to a server', (t) => {
...
+test('close with async onClose handlers', (t) => {
+  t.plan(5)
+
+  const app = boot()
+  var order = [1, 2, 3, 4]
+
+  app.onClose(async () => {
+    await new Promise(resolve => setTimeout(resolve, 500))
+    t.is(order.shift(), 3)
+  })
```
Comment (mcollina) [inline test/close.test.js]: Can you use Promises instead? We still support Node 6 and Node 4.

**Comment by gyszalai** [comment]
hm... so the tests must be ES5, no async keyword

**Comment by HelloEdit** [comment]
Do we need to update Typescript definition ?

**Comment by mcollina** [comment]
Yes please
Il giorno ven 3 ago 2018 alle 20:04 HelloEdit <notifications@github.com> ha
scritto:

**Pull request merged**